### PR TITLE
Add mu opt out to MSI install

### LINKED ISSUE #6118: Add microsoft Update opt-in functionality to MSI
In the future we plan on supporting security updates through Microsoft Update, for this to work well, the installer should ask users to opt-in to Microsoft Update.

## Workflow

- [ ] check existing MU opt-in status
    - [ ] If already opt-in to MU, go to the last step
    - [ ] if not, go to next step.
- [x] Show opt-in UI
    - [x] if the user opts out, go to the last step
    - [x] if the user opts in, got to next step
- [x] Add MU service using WUA API (see examples in a separate section)
- [ ] Configure AU
    - [ ]  If Windows 10, goto to last step
    - [ ] if AU is on, goto last step
    - [ ] check OS install type
        - [ ] if client OS set AU notification level (see examples in a separate section) to `scheduled installs`
        - [ ] if server OS set AU notification level (see examples in a separate section) to `notify before installation` 
- [ ] Continue with the installation
    
### PowerShell
https://github.com/PowerShell/xWindowsUpdate/blob/dev/DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1#L490

### API

https://msdn.microsoft.com/en-us/library/windows/desktop/aa385824(v=vs.85).aspx

### in PowerShell

https://github.com/dsccommunity/xWindowsUpdate/blob/6aab96e94e4a53b817e9f9b7cc403b8036c8b93f/source/DSCResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1#L747

### VB

And the documentation:  https://msdn.microsoft.com/en-us/library/windows/desktop/aa826676(v=vs.85).aspx

### example
Here is what the screen should say and look like (icon blurred until we get permission to use them):
image

### text
Text | Location | Comments
-- | -- | --
Use Microsoft Update to help keep your computer secure and up to date | Header |  
Microsoft Update helps keep your computer secure and up to date for Windows and other Microsoft products, including PowerShell 7. Updates will be delivered based on your current update settings. You can review or change these settings from the Windows Update control panel. | Description text… top /  middle |  
Use Microsoft Update when I check for  updates (recommended) | First radio selection | Boldface
I don’t want to use Microsoft Update. | Second radio selection |  
See the Microsoft Update FAQ | Bottom left (aligned left), first link. | Link (see below for  details)
Read the Microsoft Update Privacy  Statement | Bottom left (aligned left), second  link | Link (see below for  details)

### urls
* Microsoft Update online FAQ URL 

http://go.microsoft.com/fwlink/?LinkID=57191​

* Microsoft Update privacy statement URL 

http://go.microsoft.com/fwlink/?LinkID=57190

### PR DESCRIPTION
**Pull request by TravisEz13**
# PR Summary

image

FAQ link goes to https://aka.ms/PowerShell-Microsoft-Update-FAQ
privacy link goes to: https://aka.ms/PowerShell-Microsoft-Update-Privacy-Statement 

I will make the warning about needing MU or WSUS dynamic in a later PR.  the draft, #15776, is already open for this, but we want to get the USE_MU code in as soon as possible.

## PR Context

resolves #6118
https://github.com/MicrosoftDocs/PowerShell-Docs/issues/7787

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: assets/wix/Product.wxs [hunk 1]
```diff
@@ -84,6 +84,16 @@
                   Execute="deferred"
                   Return="ignore"
                   Impersonate="no" />
+    <SetProperty Id="EnableMU"
+                 Before="EnableMU"
+                 Sequence="execute"
+                 Value="&quot;[VersionFolder]pwsh.exe&quot; -NoProfile -ExecutionPolicy Bypass -Command &quot;$null=(New-Object -ComObject Microsoft.Update.ServiceManager).AddService2('7971f918-a847-4430-9279-4a52d1efe18d', 7, '')&quot;" />
```
Comment (iSazonov) [inline assets/wix/Product.wxs:90]: What if it is first pwsh install (no pwsh on the computer)?

Comment (TravisEz13) [reply assets/wix/Product.wxs:90]: This happens after `pwsh` is installed like all the other actions.

File: assets/wix/Product.wxs [hunk 2]
```diff
@@ -317,6 +338,36 @@
...
+      <Control Id="BannerLine" Type="Line" X="0" Y="44" Width="370" Height="0" TabSkip="yes" />
+      <Control Id="Description" Type="Text" X="25" Y="54" Width="280" Height="100" Transparent="yes" NoPrefix="yes" Text="Microsoft Update helps keep your computer secure and up to date for Windows and other Microsoft products, including PowerShell 7. Updates will be delivered based on your current update settings. You can review or change these settings from the Windows Update control panel." TabSkip="yes" />
+      <!-- If the checkboxes are defined first, then they are first in the tab order and can be ticked and unticked using the spacebar -->
+      <Control Id="UseMuCheckBox" Type="CheckBox" X="20" Y="100" Width="290" Height="27" Property="USE_MU" CheckBoxValue="0" Text="Enable updating PowerShell through Microsft Update or WSUS (recommended)"/>
+      <Control Id="EnableMuCheckBox" Type="CheckBox" X="20" Y="130" Width="290" Height="17" Property="ENABLE_MU" CheckBoxValue="0" Text="Use Microsoft Update when I check for updates (recommended)"/>
```
Comment (SteveL-MSFT) [inline assets/wix/Product.wxs:354]: The text from the Windows Update settings uses:

> Receive updates for other Microsoft products when you update Windows

We should use the same language.

File: test/tools/Modules/HelpersCommon/HelpersCommon.psm1 [hunk 3]
```diff
@@ -219,22 +219,21 @@ function Send-VstsLogFile {
...
-        $name = Split-Path -Leaf -Path $logFile
+        $logFile = Join-Path -Path $logFolder -ChildPath $newName
 
         $Contents | Out-File -path $logFile -Encoding ascii
     }
     else
     {
-        $name = Split-Path -Leaf -Path $path
-        $logFile = Join-Path -Path $logFolder -ChildPath ([System.Io.Path]::GetRandomFileName() + '-' + $name)
+        $logFile = Join-Path -Path $logFolder -ChildPath $newName
```
Comment (SteveL-MSFT) [inline test/tools/Modules/HelpersCommon/HelpersCommon.psm1:231]: Why was this change needed?  I the old code, the two paths would result in different random files, but now would have the same.

File: assets/wix/Product.wxs [hunk 4]
```diff
@@ -317,6 +338,36 @@
...
+      <Control Id="Title" Type="Text" X="15" Y="6" Width="200" Height="30" Transparent="yes" NoPrefix="yes" Text="{\WixUI_Font_Title}Use Microsoft Update to help keep your computer secure and up to date" TabSkip="yes" />
+      <Control Id="BannerLine" Type="Line" X="0" Y="44" Width="370" Height="0" TabSkip="yes" />
+      <Control Id="Description" Type="Text" X="25" Y="54" Width="280" Height="100" Transparent="yes" NoPrefix="yes" Text="Microsoft Update helps keep your computer secure and up to date for Windows and other Microsoft products, including PowerShell 7. Updates will be delivered based on your current update settings. You can review or change these settings from the Windows Update control panel." TabSkip="yes" />
+      <!-- If the checkboxes are defined first, then they are first in the tab order and can be ticked and unticked using the spacebar -->
+      <Control Id="UseMuCheckBox" Type="CheckBox" X="20" Y="100" Width="290" Height="27" Property="USE_MU" CheckBoxValue="0" Text="Enable updating PowerShell through Microsft Update or WSUS (recommended)"/>
```
Comment (SteveL-MSFT) [inline assets/wix/Product.wxs:353]: Can you have an action here where if this box is checked, the EnableMuCheckBox is also checked since this one depends on the other?  Also, because of the dependency, should the order of these two boxes be switched?

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

File: assets/wix/Product.wxs (hunk 2 above)
Comment (TravisEz13) [reply assets/wix/Product.wxs:354]: This does not meet the spec.

File: assets/wix/Product.wxs (hunk 4 above)
Comment (TravisEz13) [reply assets/wix/Product.wxs:353]: The action has to be in the actions, putting it here is not valid syntax.  GUI cannot perform action.

I can switch the order.

File: assets/wix/Product.wxs [hunk 5]
```diff
@@ -317,6 +338,36 @@
...
+      <Control Id="EnableMuCheckBox" Type="CheckBox" X="20" Y="130" Width="290" Height="17" Property="ENABLE_MU" CheckBoxValue="0" Text="Use Microsoft Update when I check for updates (recommended)"/>
+      <Control Id="MuFAQ" Type="Hyperlink" X="20" Y="190" Width="214" Height="17">
+        <Text><![CDATA[<a href="https://aka.ms/PowerShell-Microsoft-Update-FAQ">See the Microsoft Update FAQ.</a>]]></Text>
```
Comment (anamnavi) [inline assets/wix/Product.wxs:356]: This link takes me to a BING homepage, doesn't contain any Microsoft Update information (as far as I can tell)

Comment (TravisEz13) [reply assets/wix/Product.wxs:356]: We'll need to update the link, but this can be done after the get the code committed.

File: test/tools/Modules/HelpersCommon/HelpersCommon.psm1 (hunk 3 above)
Comment (TravisEz13) [reply test/tools/Modules/HelpersCommon/HelpersCommon.psm1:231]: The code was changing the name when uploading but not logging how it changed the name.  I just changed it to log what it changed the name to.

**Review by anamnavi — APPROVED** [review]

File: assets/wix/Product.wxs (hunk 5 above)
Comment (TravisEz13) [reply assets/wix/Product.wxs:356]: Filed a doc request for this:  https://github.com/MicrosoftDocs/PowerShell-Docs/issues/7787

**Review by AungMinKo070 — APPROVED** [review]

**Comment by TravisEz13** [comment]
remove periods from end of hyperlinks

**Comment by TravisEz13** [comment]
add warning like this
```
Warning: Without Microsoft Update enabled,
you will need to use another update solution like WSUS or SCCM
in order to receive automatic updates.
Are you sure you want to continue?



Okay / Cancel
```

**Review by aungminko93750 — APPROVED** [review]

File: assets/wix/Product.wxs [hunk 6]
```diff
@@ -343,6 +396,8 @@
 
       <Property Id="DefaultUIFont" Value="WixUI_Font_Normal" />
       <Property Id="WixUI_Mode" Value="InstallDir" />
+      <PropertyRef Id="ENABLE_MU" />
+      <PropertyRef Id="USE_MU" />
```
Comment (TravisEz13) [inline assets/wix/Product.wxs:400]: remove dead code

**Review by adityapatwardhan — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
@SteveL-MSFT  can you update your review?

**Review by SteveL-MSFT — APPROVED** [review]
